src/core/tranche/slicer.sol

```diff
@@ -0,0 +1,145 @@
+// Copyright (C) 2019  Centrifuge
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.4.24;
+
+import "ds-note/note.sol";
+
+// Slicer
+//  Calculates payouts and slices based on the current iSupply chi. Manages iSupply.
+contract Slicer is DSNote {
+    // --- Auth ---
+    mapping (address => uint) public wards;
+    function rely(address usr) public auth note { wards[usr] = 1; }
+    function deny(address usr) public auth note { wards[usr] = 0; }
+    modifier auth { require(wards[msg.sender] == 1); _; }
+
+    struct Fee {
+        uint chi;
+        uint speed;
+        uint48 rho;
+    }
+    
+    Fee public iSupply;
+    
+    constructor() public {
+        wards[msg.sender] = 1;
+        iSupply.chi = ONE;
+        iSupply.speed = ONE;
+    }
+
+    function file(bytes32 what, uint speed_) public note auth {
+        if (what == "isupply") { 
+            drip();
+            iSupply.speed = speed_;    
+        } 
+    }
+
+    function calcSlice(uint currencyAmount) public note auth returns (uint) {
+        if (now > iSupply.rho) {
+            drip();
+        }
+        uint slice = rdiv(currencyAmount, iSupply.chi);
+        return slice;
+    }
+
+    function calcPayout(uint tokenAmount) public note auth returns (uint) {
+        if (now > iSupply.rho) {
+            drip();
+        }
+        uint payout = rmul(tokenAmount, iSupply.chi);
+        return payout;
+    }
+
+    function drip() public note auth {
+        if (now >= iSupply.rho) {
+            (uint latest,) = compounding();
+            iSupply.chi = latest;
+            iSupply.rho = uint48(now);
+        }
+    }
+
+    function compounding() public view returns (uint, uint) {
+        uint48 rho = iSupply.rho;
+        require(now >= rho);
+        uint speed = iSupply.speed;
+        uint chi = iSupply.chi;
+        // compounding in seconds
+        uint latest = rmul(rpow(speed, now - rho, ONE), chi);
+        uint chi_ = rdiv(latest, chi);
+        return (latest, chi_);
+    }
+
+    function updateISupply(uint borrowSpeed, uint debt, uint reserve) public note auth {
+        require (borrowSpeed > 0);
+        if (now >= iSupply.rho) {
+            drip();
+        }
```

Comment:
here the question is if we should call the `drip` after the new `supplyRate` (iSupply.speed)?

We find a system state and we need to calc the interest let's say for the last one hour.

Should we use the current system state for it or the system state from one hour ago ?